apollo-spaceport/src/server.rs

```diff
@@ -0,0 +1,330 @@
+use crate::{
+    agent::{
+        reporter_server::{Reporter, ReporterServer},
+        ReporterGraph, ReporterResponse,
+    },
+    report::Report,
+    ReporterStats, ReporterTrace, TracesAndStats,
+};
+use bytes::BytesMut;
+use flate2::write::GzEncoder;
+use flate2::Compression;
+use prost::Message;
+use prost_types::Timestamp;
+use reqwest::Client;
+use std::collections::HashMap;
+use std::io::Write;
+use std::net::SocketAddr;
+use std::sync::atomic::{AtomicU32, Ordering};
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
+use tokio::sync::mpsc::Sender;
+use tokio::sync::Mutex;
+use tokio::time::{interval, Duration, MissedTickBehavior};
+use tonic::transport::{Error, Server};
+use tonic::{Request, Response, Status};
+
+type TPQMap = HashMap<String, TracesAndStats>;
+type GraphMap = Arc<Mutex<HashMap<ReporterGraph, TPQMap>>>;
+
+static DEFAULT_INGRESS: &str = "https://usage-reporting.api.apollographql.com/api/ingress/traces";
+static INGRESS_CLOCK_TICK: Duration = Duration::from_secs(5);
+static TRIGGER_BATCH_LIMIT: u32 = 50;
+
+/// Allows common transfer code to be more easily represented
+#[allow(clippy::large_enum_variant)]
+enum StatsOrTrace {
+    Stats(ReporterStats),
+    Trace(ReporterTrace),
+}
+
+impl StatsOrTrace {
+    fn get_traces_and_stats(&self) -> TracesAndStats {
+        match self {
+            StatsOrTrace::Stats(_) => TracesAndStats {
+                stats_with_context: vec![],
+                ..Default::default()
+            },
+            StatsOrTrace::Trace(_) => TracesAndStats {
+                trace: vec![],
+                ..Default::default()
+            },
+        }
+    }
+
+    fn graph(&self) -> Option<ReporterGraph> {
+        match self {
+            StatsOrTrace::Stats(s) => s.graph.clone(),
+            StatsOrTrace::Trace(t) => t.graph.clone(),
+        }
+    }
+
+    fn key(&self) -> String {
+        match self {
+            StatsOrTrace::Stats(s) => s.key.clone(),
+            StatsOrTrace::Trace(t) => t.key.clone(),
+        }
+    }
+}
+
+/// Accept Traces and Stats from clients and transfer to an Apollo Ingress
+pub struct ReportSpaceport {
+    addr: SocketAddr,
+    // This Map will only have a single entry if used internally from a router.
+    // (because a router can only be serving a single graph)
+    tpq: GraphMap,
```

Comment:
what does tpq mean here?

Reply:
It was taken from the Report protobuf definition: traces_per_query.

It's not a great name, so I thought about what these maps actually represent and did some re-naming to make it clearer. tpq is now graph_usage and I've propagated that naming theme throughout the code. I hope it makes the code a bit clearer.

Reply:
done: 2b01014ee6048ace2fd7cce3b679cc87fc60c752